share/tunnel/tunnel.go

```diff
@@ -174,16 +174,42 @@ func (t *Tunnel) BindRemotes(ctx context.Context, remotes []*settings.Remote) er
 
 func (t *Tunnel) keepAliveLoop(sshConn ssh.Conn) {
 	//ping forever
+
+	//reply_timeout set to KeepAlive interval, if KeepAlive is less than 10s, set reply_timeout to 10s
+	//maybe a new config option for reply_timeout is also fine
+	reply_timeout := t.Config.KeepAlive
+
+	if reply_timeout < 10*time.Second {
+		reply_timeout = 10 * time.Second
+	}
+
 	for {
 		time.Sleep(t.Config.KeepAlive)
-		_, b, err := sshConn.SendRequest("ping", true, nil)
+
+		errChannel := make(chan error, 2)
+
+		time.AfterFunc(reply_timeout, func() {
+			errChannel <- errors.New("KEEPALIVE REPLY TIMEOUT ERROR")
+		})
+
+		go func() {
+			_, b, err := sshConn.SendRequest("ping", true, nil)
+
```

Comment:
I believe you should explicitly cancel the timeout after receiving a response from "ping" request. It wouldn't make a difference on the behavior just ensuring the resource is explicitly cleared.

Reply:
This has been added in the last commit 42125fed3dfb5dfc154b7fbacce733be20e5faef